src/runtime/opencl.cpp

```diff
@@ -936,30 +956,69 @@ WEAK int halide_opencl_run(void *user_context,
 
     // Set args
     int i = 0;
+
+    // Count sub buffers needed for crops.
+    int sub_buffers_needed = 0;
+    while (arg_sizes[i] != 0) {
+        if (arg_is_buffer[i] &&
+            ((device_handle *)((halide_buffer_t *)args[i])->device)->offset != 0) {
+            sub_buffers_needed++;
+        }
+        i += 1;
+    }
+    cl_mem *sub_buffers = NULL;
+    int sub_buffers_saved = 0;
+    if (sub_buffers_needed > 0) {
+        sub_buffers = (cl_mem *)malloc(sizeof(cl_mem) * sub_buffers_needed);
+        if (sub_buffers == NULL) {
+            return halide_error_code_out_of_memory;
+        }
+        memset(sub_buffers, 0, sizeof(cl_mem) * sub_buffers_needed);
+    }
+
+    i = 0;
     while (arg_sizes[i] != 0) {
         debug(user_context) << "    clSetKernelArg " << i
                             << " " << (int)arg_sizes[i]
                             << " [" << (*((void **)args[i])) << " ...] "
                             << arg_is_buffer[i] << "\n";
         void *this_arg = args[i];
-        cl_int err;
+        cl_int err = CL_SUCCESS;
 
         if (arg_is_buffer[i]) {
             halide_assert(user_context, arg_sizes[i] == sizeof(uint64_t));
-            uint64_t opencl_handle = ((halide_buffer_t *)this_arg)->device;
-            debug(user_context) << "Mapped dev handle is: " << (void *)opencl_handle << "\n";
-            // In 32-bit mode, opencl only wants the bottom 32 bits of
-            // the handle, so use sizeof(void *) instead of
-            // arg_sizes[i] below.
-            err = clSetKernelArg(f, i, sizeof(void *), &opencl_handle);
+            cl_mem mem = ((device_handle *)((halide_buffer_t *)this_arg)->device)->mem;
+            uint64_t offset = ((device_handle *)((halide_buffer_t *)this_arg)->device)->offset;
+
+            uint64_t opencl_handle;
+            if (offset == 0) {
```

Comment:
It would also make sense to check the size here. A crop might have identical mins, but smaller extents.

Reply:
In order to do that I think I'm going to have to ask OpenCL for the size of the cl_mem via a cal to clGetMemObjectInfo. This would happen for each cl_mem argument that is not a crop. (As well as those that are crops with the same mins.) Currently we only call clGetMemObjectInfo in the debug path I believe. It strikes me as both likely to be inefficient and might result in an unnecessary failures somewhere.

The only advantage would be possibly better bounds checking in the OpenCL implementation if it is even doing bounds checking, but we should never have that sort of error in Halide unless someone lies about the size of a buffer when wrapping it. (Which I could see protecting against by an assert in the wrap function which guarantees the cl_mem is large enough for the bounds passed in.)

Reply:
Hmm, yeah. FWIW, my motivation for this was not bounds checking, but assuming that OpenCL might be able to optimize things if it knows the memory used is smaller than the buffer itself.